SignalServiceKit/src/Contacts/TSThread.h

```diff
@@ -157,18 +157,19 @@ NS_ASSUME_NONNULL_BEGIN
 @property (atomic, readonly, nullable) NSDate *mutedUntilDate;
 
 // This model may be updated from many threads. We don't want to save
-// our local copy (this instance) since it may be out of date.  Instead, we
+// our local copy (this instance) since it may be out of date.  We also
+// want to avoid re-saving a model that has been deleted.  Therefore, we
```

Comment:
I've made parallel changes in `TSOutgoingMessage` and `TSThread`.  These changes are far more impactful for `TSOutgoingMessage`, but I'll explain once here.

The `updateWith...` approach has been fairly successful in avoiding the class of bugs where stale data is saved to the database, but as initially implemented it had the undesirable risk of re-saving deleted models.  

I'm changing the semantics of these methods to only save if there already is an existing copy of the model in the database.

In a number of places, we were relying on this side-effect of `updateWith...` methods and using them to save new models.  This was never a good idea.  We now are explicit about when to save models to the database. In the case of outgoing messages, we save them:

* When they are enqueued in the message sender.
* When we learn of an outgoing message from a linked device via transcript.

Although an important change, I believe this should be safe since outgoing messages should only be saved at these two points.